service/automatic_run_experiment.py

```diff
@@ -62,26 +73,128 @@
+
+
+def validate_experiment_name(experiment_name):
+    """Returns True if |experiment_name| is valid."""
+    return EXPERIMENT_NAME_REGEX.match(experiment_name) is not None
+
+
+def validate_experiment_requests(experiment_requests):
+    """Returns True if all requests in |experiment_requests| are valid."""
+    # This function tries to find as many requests as possible.
+    if PAUSE_SERVICE_KEYWORD in experiment_requests:
+        # This is a special case where a string is used instead of an experiment
+        # to tell the service not to run experiments automatically. Remove it
+        # from the list because it fails validation.
+        experiment_requests = experiment_requests[:]  # Don't mutate input.
+        experiment_requests.remove(PAUSE_SERVICE_KEYWORD)
+
+    all_fuzzers = set(fuzzer_utils.get_fuzzer_names())
+    valid = True
+    # Validate format.
+    for request in experiment_requests:
+        if not isinstance(request, dict):
+            logger.error('Request: %s is not a dict.', request)
+            experiment_requests.remove(request)
+            valid = False
+            continue
+
+        if 'experiment' not in request:
+            logger.error('Request: %s does not have field "experiment".',
+                         request)
+            valid = False
+            continue
+
+        experiment = _get_experiment_name(request)
+        if not validate_experiment_name(experiment):
+            valid = False
+            logger.error('Experiment name: %s is not valid.', experiment)
+            # Request isn't so malformed that we can finding other issues.
+            # if present. Don't continue.
+
+        experiment = request['experiment']
+        if not request.get('fuzzers'):
+            logger.error('Request: %s does not specify any fuzzers.', request)
+            valid = False
+            continue
+
+        experiment_fuzzers = request['fuzzers']
+        for fuzzer in experiment_fuzzers:
+            if fuzzer in all_fuzzers:
+                continue
+            # Fuzzer isn't valid.
+            logger.error('Fuzzer: %s in experiment %s is not valid.', fuzzer,
+                         experiment)
+            valid = False
+
+    if not valid:
+        # Don't try the next validation step if the previous failed, we might
+        # exception.
+        return valid
+
+    # Make sure experiment requests have a unique name, we can't run the same
+    # experiment twice.
+    counts = collections.Counter(
+        [request['experiment'] for request in experiment_requests])
+
+    for experiment_name, count in counts.items():
+        if count != 1:
+            logger.error('Experiment: "%s" appears %d times.',
+                         str(experiment_name), count)
+            valid = False
+
+    return valid
+
+
+def run_requested_experiment(dry_run):
+    """Run the oldest requested experiment that hasn't been run yet in
+    experiment-requests.yaml."""
```

Comment:
This means we will run one request a day? If multiple requests are made one day, would it make sense to merge them?

Reply:
It probably would. There are definitely some problems with the current approach that should be fixed (I assume many pull requests won't even be merged on the day they were worked on, so I'd expect with this system well run experiments named 2020-01-01 on 2020-01-02 etc.)